packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -113,108 +114,111 @@ const withInfiniteScroll = (
+              elementsToRender: nextElements,
+              showSeeMore: seeMoreAfter === index / elementsPerScroll,
+            };
+          }
+          if (onScrollFinished) {
+            onScrollFinished({
+              totalNumberElements: elementsToRender.length,
+            });
+          }
           return {
-            index: nextIndex,
-            elementsToRender: elementsToRender.concat(nextElements),
-            showSeeMore: seeMoreAfter === index / elementsPerScroll,
+            isListFinished: true,
           };
-        }
-        if (onScrollFinished) {
-          onScrollFinished({
-            totalNumberElements: elementsToRender.length,
-          });
-        }
-        return {
-          isListFinished: true,
-        };
-      });
+        });
     }
 
-    handleIntersection = (entries: Array<IntersectionObserverEntry>) => {
+    handleIntersection = () => {
       const { onScroll } = this.props;
-      const entry = entries[0];
-      if (entry.intersectionRatio >= 1) {
-        if (this.sentinel) {
-          this.observer.unobserve(this.sentinel);
-        }
-        if (onScroll) {
-          onScroll({ currentIndex: this.state.index });
-        }
-        return this.loadMore().then(newState => {
-          this.setState(newState);
-        });
+      if (onScroll) {
+        onScroll({ currentIndex: this.state.index });
       }
-      return Promise.resolve();
+      this.loadMore().then(newState => {
+        this.setState(newState);
+      });
     };
 
-    handleSeeMoreClick = () =>
+    handleSeeMoreClick = () => {
       this.loadMore().then(newState => {
         this.setState(newState);
       });
+    };
 
     render() {
-      const { elementsToRender, isListFinished, showSeeMore } = this.state;
+      const {
+        elementsToRender,
+        isListFinished,
+        showSeeMore,
+        index,
+      } = this.state;
       const { renderSeeMoreComponent, renderLoadingComponent } = this.props;
 
       const rest = omit(this.props, Object.keys(propTypes));
 
-      let loadingOrButton = null;
+      let LoadingComponent = null;
+      let seeMoreButton = null;
 
       if (!isListFinished) {
         if (showSeeMore && renderSeeMoreComponent) {
-          loadingOrButton = renderSeeMoreComponent({
+          seeMoreButton = renderSeeMoreComponent({
             onSeeMoreClick: this.handleSeeMoreClick,
           });
         } else {
-          loadingOrButton = (
+          LoadingComponent = ({ onRef }) => (
             <div
-              ref={spinner => {
-                this.sentinel = spinner;
-              }}
+              ref={onRef}
               className={
                 renderLoadingComponent ? null : getClassNames('bpk-sentinel')
               }
             >
               {renderLoadingComponent && renderLoadingComponent()}
             </div>
           );
+          if (index > 0) {
+            LoadingComponent = withIntersectionObserver(LoadingComponent);
```

Comment:
This is not a very good idea performance wise. We are recreating this HOC at every render. Is there a problem with the previous implementation?

Reply:
not a problem just wanted to abstract it